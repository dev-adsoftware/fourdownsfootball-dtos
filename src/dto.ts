import { ClassConstructor, plainToClass } from 'class-transformer';
import { validateSync, ValidationError } from 'class-validator';

export abstract class Dto {
  public init(obj: Record<string | number, unknown>): this {
    Object.assign(
      this,
      plainToClass(this.constructor as ClassConstructor<this>, { ...obj }),
    );
    return this.validate(true);
  }

  public serialize(): string {
    return JSON.stringify(this);
  }

  private validate(whitelist: boolean): this {
    const errors = validateSync(this, {
      whitelist,
      validationError: { target: false },
    });
    if (errors.length > 0) {
      throw new Error(
        `${this.constructor.name}: ${this.validationErrors(errors)}`,
      );
    }

    return this;
  }

  private validationErrors(errors: ValidationError[]): string {
    return errors
      .map((e) => {
        if (e.constraints) return Object.values(e.constraints).join(', ');
        return String(e);
      })
      .join(', ');
  }
}
