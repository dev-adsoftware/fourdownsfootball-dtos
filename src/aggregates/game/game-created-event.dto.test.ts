import { createDto } from '../../dto';
import { GameCreatedEvent } from '../..';

describe('given: no game created event data', () => {
  describe('when: I create an game created event dto', () => {
    it('then: dto threw expected error', async () => {
      expect(() =>
        createDto(GameCreatedEvent, {}),
      ).toThrowErrorMatchingSnapshot();
    });
  });
});

describe('given: proper game', () => {
  describe('when: I create an game dto', () => {
    it('then: dto was created with expected values', async () => {
      expect(
        createDto(GameCreatedEvent, {
          homeUsername: 'jest.homeUsername',
          homeTeamId: 'jest.homeTeamId',
          awayUsername: 'jest.awayUsername',
          awayTeamId: 'jest.awayTeamId',
        }),
      ).toEqual({
        awayTeamId: 'jest.awayTeamId',
        awayUsername: 'jest.awayUsername',
        homeTeamId: 'jest.homeTeamId',
        homeUsername: 'jest.homeUsername',
      });
    });
  });
});
