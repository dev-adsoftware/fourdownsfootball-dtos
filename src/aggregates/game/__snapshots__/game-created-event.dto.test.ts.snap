// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given: no game created event data when: I create an game created event dto then: dto threw expected error 1`] = `"GameCreatedEvent: homeUsername must be a string, homeTeamId must be a string, awayUsername must be a string, awayTeamId must be a string"`;
